navigation/navigation-compose/src/androidTest/java/androidx/navigation/compose/NavBackStackEntryProviderTest.kt

```diff
@@ -0,0 +1,159 @@
+    }
+
+    @Test
+    fun testLifecycleOwnerProvided() {
+        val testNavigator = TestNavigator()
+        val navigatorState = TestNavigatorState()
+        val backStackEntry =
+            navigatorState.createActiveBackStackEntry(testNavigator.createDestination())
+        var lifecycleOwner: LifecycleOwner? = null
+
+        composeTestRule.setContent {
+            val saveableStateHolder = rememberSaveableStateHolder()
+            backStackEntry.provideToCompositionLocals(saveableStateHolder) {
+                lifecycleOwner = LocalLifecycleOwner.current
+            }
+        }
+
+        assertWithMessage("LifecycleOwner is provided by $backStackEntry")
+            .that(lifecycleOwner).isEqualTo(backStackEntry)
+    }
+
+    @Test
+    fun testLocalSavedStateRegistryOwnerProvided() {
+        val testNavigator = TestNavigator()
+        val navigatorState = TestNavigatorState()
+        val backStackEntry =
+            navigatorState.createActiveBackStackEntry(testNavigator.createDestination())
+        var localSavedStateRegistryOwner: SavedStateRegistryOwner? = null
+
+        composeTestRule.setContent {
+            val saveableStateHolder = rememberSaveableStateHolder()
+            backStackEntry.provideToCompositionLocals(saveableStateHolder) {
+                localSavedStateRegistryOwner = LocalSavedStateRegistryOwner.current
+            }
+        }
+
+        assertWithMessage("LocalSavedStateRegistryOwner is provided by $backStackEntry")
+            .that(localSavedStateRegistryOwner).isEqualTo(backStackEntry)
+    }
+
+    @Test
+    fun testSaveableValueInContentIsSaved() {
+        val restorationTester = StateRestorationTester(composeTestRule)
+        var array: IntArray? = null
+
+        val testNavigator = TestNavigator()
+        val navigatorState = TestNavigatorState()
+        val backStackEntry =
+            navigatorState.createActiveBackStackEntry(testNavigator.createDestination())
+
+        restorationTester.setContent {
+            val saveableStateHolder = rememberSaveableStateHolder()
+            backStackEntry.provideToCompositionLocals(saveableStateHolder) {
+                array = rememberSaveable {
+                    intArrayOf(0)
+                }
+            }
+        }
+
+        assertThat(array).isEqualTo(intArrayOf(0))
+
+        composeTestRule.runOnUiThread {
+            array!![0] = 1
+            // we null it to ensure recomposition happened
+            array = null
+        }
+
+        restorationTester.emulateSavedInstanceStateRestore()
+
+        assertThat(array).isEqualTo(intArrayOf(1))
+    }
+
+    // By default, NavBackStackEntrys are in the INITIALIZED state and then get moved to the next
+    // appropriate state by the NavController. In case we aren't testing with a NavController,
+    // this sets the entry's lifecycle state to the passed state so that the entry is active.
+    private fun NavigatorState.createActiveBackStackEntry(
```

Comment:
With [aosp/1697987](https://android-review.googlesource.com/c/platform/frameworks/support/+/1697987), I think you can refactor this to:
1) Create the `TestNavigator`
2) Attach a `TestNavigatorState`
3) Create a `NavBackStackEntry`
4) `navigate()` to that `NavBackStackEntry`
5) Return the created `NavBackStackEntry` (which will now be in the right state).

Reply:
omg yes, awesome, that's exactly what we need here! Thanks!

Reply:
I rewrote the tests to follow that so we'll just merge after your CL has been merged; they're obvs failing rn

Reply:
Alright - tests are passing! We're good to go once you're happy :)